### PR TITLE
Chinese yoyo

### PR DESCRIPTION
**Pull request by RussTedrake**
adds the hybrid dynamics for this special case of the tension cable system.

http://groups.csail.mit.edu/locomotion/drake/movies/chinese_yoyo.swf
http://groups.csail.mit.edu/locomotion/drake/movies/chinese_yoyo_bounce.swf

also adds some energy tools to RBM (useful for testing the hybrid dynamics) and adds the correction term for driving phi to zero which was always in my derivation, but somehow never made it into the constraint solver.

### CODE REVIEW

File: examples/SimplePulleys/ChineseYoYo.m [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  properties
+    in_contact
+    no_contact
+  end
+  
+  methods
+    function obj = ChineseYoYo
+      in_contact = PlanarRigidBodyManipulator('tension.urdf');
+
+      in_contact = PlanarRigidBodyManipulator('tension.urdf');
```
Comment (mposa) [inline examples/SimplePulleys/ChineseYoYo.m]: Repeated line here.

File: examples/SimplePulleys/ChineseYoYo.m [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      % manually remove the ball from the pulley system:
+      pulley_constraint = in_contact.position_constraints{1};
+      cable_length_fcn = pulley_constraint.fcn;
+      cable_length_fcn.pulley(2)=[];
+      pulley_constraint = DrakeFunctionConstraint(pulley_constraint.lb, pulley_constraint.ub, cable_length_fcn);
+      no_contact = in_contact.updatePositionEqualityConstraint(1,pulley_constraint);
+      
+      obj = obj@HybridDrakeSystem(0,1+getNumOutputs(in_contact));
+      obj = setInputFrame(obj,getInputFrame(in_contact));
+      obj = setOutputFrame(obj,MultiCoordinateFrame.constructFrame({CoordinateFrame('mode',1,'m',{'m'}),getOutputFrame(in_contact)}));
```
Comment (mposa) [inline examples/SimplePulleys/ChineseYoYo.m]: A comment on this frame construction would be helpful, since I'm not sure what it's doing (and why)

File: examples/SimplePulleys/ChineseYoYo.m [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    function [g,dg] = collisionGuard(obj,t,x,u)
+      % todo: generalize this using line segements of the cable constraint?
+      g = x(3) - 4;  % x(3) = load_z
+      dg = [0,0,0,1,0,0,0];
+    end
+    
+    function [g,dg] = aerialGuard(obj,t,x,u)
+      g = 4 - x(3);
+      dg = -[0,0,0,1,0,0,0];
+    end
```
Comment (mposa) [inline examples/SimplePulleys/ChineseYoYo.m]: What are these two guard functions computing?

File: examples/SimplePulleys/ChineseYoYo.m [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        modep = 1;
+        
+        % note: this shouldn't do any mechanical work (phidot should already 
+        % be zero at the transition, because the in_contact phi is equivalent 
+        % to the no_contact phi when the ball is tangent to the cable), but 
+        % let's put it in for numerical reasons
+        [phi,J] = obj.no_contact.position_constraints{1}.eval(q);
+      end
+      % solve zero post-transition velocities for all active constraints
+      J = J*vToqdot(obj.no_contact, q);
```
Comment (mposa) [inline examples/SimplePulleys/ChineseYoYo.m]: Should this use obj.in_contact or obj.no_contact, depending on mode?

**Comment by RussTedrake** [comment]
awesome.  responded to addressed them all.  thanks.
